File: day3/day3_2.py
def most_common_n(n, list):
    ret = ['0','1']
    freq1, freq0 = 0, 0
    for i in list:
        if i[1][n] == '1':
            freq1 += 1
        else:
            freq0 += 1
    return ret[freq1 >= freq0]

def reduce_to_largest(list):
    list = [*enumerate(list)]
    i = 0
    l = len(list[0][1])
    while (len(list)>1 and i < l):
        list = [(x,y) for (x,y) in list if y[i] == most_common_n(i, list)]
        i += 1
    return list[0]

def reduce_to_smallest(list):
    list = [*enumerate(list)]
    i = 0
    l = len(list[0][1])
    while (len(list)>1 and i < l):
        list = [(x,y) for (x,y) in list if y[i] != most_common_n(i, list)]
        i += 1
    return list[0]

with open ("3_1.txt", "r") as problem:
    lines = problem.readlines()
    print (len(lines[0]))
    print(reduce_to_smallest(lines))
    print(reduce_to_largest(lines))

File: day3/day3_2.clj
(require '[clojure.string :as strn])

(def input (-> "3_1.txt"
               slurp
               (strn/split-lines)))

(defn most-common-n [n coll]
  (let [chars (map #(get-in % [1 n]) coll)
        freq0 (reduce (fn [acc item] (if (= \0 item) (inc acc) acc)) 0 chars)
        freq1 (- (count chars) freq0)]
    (if (> freq0 freq1) \0 \1)
    ))

(defn reduce-to-largest-and-smallest [coll]
  (let [i-list (range (count coll))
        partitions (group-by (fn [[i y]] (= (most-common-n i coll) (get y i)))
                             (map (fn [a b] [a b]) i-list coll))]
    [(first (get partitions true)), (first (get partitions false))]))

(let [[largest smallest] input]
  (println smallest)
  (println largest))
